pkg/cloudprovider/provider/vsphere/provider.go

```diff
@@ -0,0 +1,402 @@
+		}
+	}
+
+	_, err = finder.Datastore(context.TODO(), config.Datastore)
+	if err != nil {
+		return cloudprovidererrors.TerminalError{
+			Reason:  v1alpha1.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to get vsphere datastore %s due to %v", config.Datastore, err),
+		}
+	}
+
+	_, err = finder.ClusterComputeResource(context.TODO(), config.Cluster)
+	if err != nil {
+		return cloudprovidererrors.TerminalError{
+			Reason:  v1alpha1.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to get vsphere cluster %s due to %v", config.Cluster, err),
+		}
+	}
+	return nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, userdata string) (instance.Instance, error) {
+	config, _, err := p.getConfig(machine.Spec.ProviderConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := getClient(config.Username, config.Password, config.VSphereURL, config.AllowInsecure)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = CreateLinkClonedVm(machine.Spec.Name,
+		config.TemplateVMName,
+		config.Datacenter,
+		config.Cluster,
+		config.CPUs,
+		config.MemoryMB,
+		client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create linked vm: '%v'", err)
+	}
+
+	finder, err := getDatacenterFinder(config.Datacenter, client)
+	if err != nil {
+		return nil, err
+	}
+	virtualMachine, err := finder.VirtualMachine(context.TODO(), machine.Spec.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get virtual machine object: %v", err)
+	}
+
+	localUserdataIsoFilePath, err := generateLocalUserdataIso(userdata, machine.Spec.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		err := os.Remove(localUserdataIsoFilePath)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("failed to clean up local userdata iso file at %s: %v", localUserdataIsoFilePath, err))
+		}
+	}()
+
+	err = uploadAndAttachISO(finder, virtualMachine, localUserdataIsoFilePath, config.Datastore, client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to upload and attach userdata iso: %v", err)
+	}
+
+	// Ubuntu wont boot with attached floppy device, because it tries to write to it
+	// which fails, because the floppy device does not contain a floppy disk
+	// Upstream issue: https://bugs.launchpad.net/cloud-images/+bug/1573095
+	err = removeFloppyDevice(virtualMachine)
+	if err != nil {
+		return nil, err
+	}
+
+	powerOnTask, err := virtualMachine.PowerOn(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("failed to power on machine: %v", err)
+	}
+
+	powerOnTaskContext, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	err = powerOnTask.Wait(powerOnTaskContext)
+	if err != nil {
+		return nil, fmt.Errorf("timed out waiting to power on vm %s: %v", virtualMachine.Name(), err)
+	}
+
+	return VSphereServer{name: virtualMachine.Name(), status: instance.StatusRunning, id: virtualMachine.Reference().Value}, nil
+}
+
+func (p *provider) Delete(machine *v1alpha1.Machine) error {
```

Comment:
I've not used TerminalErrors here, because if something fails here its most likely due to the fact that something was deleted in vsphere manually. The only exceptions to that are `getConfig` and `getClient` which return TerminalErrors by themselves